File: src/core/MappingTypes.cpp
// ============================================================
// MappingTypes.cpp
// 映射数据结构的实现：描述生成 与 默认配置创建
// ------------------------------------------------------------
// 本文件包含两个部分：
//   1. KeyMapping::describe()：把一条映射翻译成人类可读的文本，
//      用于编辑对话框按钮列表/悬浮窗展示（如 "B+Ctrl"）。
//   2. ControllerProfile::createDefault()：生成一份开箱即用的默认
//      配置——一个公共层 + 10 个操作层（只带 WoW 预设显示名，
//      不预设任何按键映射，由用户编辑）。
//
// 分层模型回顾（详见 MappingTypes.h）：
//   - 公共层(Common)：始终处于激活兜底，是配置 SwitchLayer
//     （层切换）映射的常用位置（操作层内亦可设置切换层）。
//   - 操作层(Layer1~10)：由公共层里的 SwitchLayer 映射"按住激活/
//     松开回退"临时叠加，查询顺序为激活层栈顶 -> ... -> 公共层。
// ============================================================

// 【C++ 语法】#include "..."：包含本文件对应的头文件，把 MappingTypes.h 中声明的类与函数定义引入。
#include "MappingTypes.h"

// 【Qt】#include <QStringList>：尖括号形式包含 Qt 库头文件；QStringList 用于 describe() 中收集展示片段。
#include <QStringList>

// ============================================================
// KeyMapping::describe：把一条映射描述为可读文本
// ============================================================
// 格式：主动作 [+ 子命令1] [+ 子命令2] [+ 子命令3]
// 例如：键盘按键"B" + 子命令 Ctrl 会显示为 "B+Ctrl"（组合键）。
// 该文本只用于 UI 展示，与配置文件里的结构化字段无关。
// 【C++ 语法】类外定义成员函数："类名::函数名"；const 与 .h 声明保持一致，表示不修改对象。
// 返回值 QString 为 Qt 字符串（按值返回，内部隐式共享）。
QString KeyMapping::describe() const {      // 实现 const 成员函数：生成该映射的可读描述文本
    QStringList parts;                      // 收集各展示片段的字符串列表
    switch (action.type) {                  // 【C++ 语法】switch：按主动作类型分支处理
        case MappedAction::Type::KeyboardKey:    // 键盘按键：显示键名
            parts << keyCodeToName(action.keyCode);   // 【Qt】<< 追加元素；keyCodeToName() 把键码翻译成键名
            break;                                  // 【C++ 语法】break：结束当前分支，防止贯穿到下一 case
        case MappedAction::Type::MouseClick:        // 鼠标单击：显示鼠标按钮名
            parts << mouseButtonDisplayName(action.mouseButton);   // 鼠标按钮转显示名
            break;                                              // 跳出 switch
        case MappedAction::Type::MouseToggle:       // 鼠标长按锁存：显示"长按+按钮名"
            parts << QStringLiteral("长按%1").arg(mouseButtonDisplayName(action.mouseButton));   // 【Qt】arg() 把 %1 替换为按钮名
            break;                                              // 跳出 switch
        case MappedAction::Type::WheelUp:           // 滚轮上滚
            parts << QStringLiteral("滚轮上滚");    // 追加固定文本"滚轮上滚"
            break;                                  // 跳出 switch
        case MappedAction::Type::WheelDown:         // 滚轮下滚
            parts << QStringLiteral("滚轮下滚");    // 追加固定文本"滚轮下滚"
            break;                                  // 跳出 switch
        case MappedAction::Type::SwitchLayer:       // 切换操作层：显示"切换→层名"
            parts << QStringLiteral("切换→%1").arg(action.layerName);   // 【Qt】arg() 把 %1 替换为目标层名
            break;                                  // 跳出 switch
        case MappedAction::Type::MouseMove:         // 鼠标移动
            parts << QStringLiteral("鼠标移动");    // 追加固定文本"鼠标移动"
            break;                                  // 跳出 switch
        case MappedAction::Type::LookAround:        // 视角控制
            parts << QStringLiteral("视角控制");    // 追加固定文本"视角控制"
            break;                                  // 跳出 switch
        case MappedAction::Type::ToggleMapping:     // 切换映射
            parts << QStringLiteral("切换映射");    // 追加固定文本"切换映射"
            break;                                  // 跳出 switch
        case MappedAction::Type::ToggleOnScreenKeyboard:   // 切换屏幕键盘
            parts << QStringLiteral("切换屏幕键盘");        // 追加固定文本"切换屏幕键盘"
            break;                                         // 跳出 switch
        case MappedAction::Type::ToggleOverlay:     // 切换悬浮窗
            parts << QStringLiteral("切换悬浮窗");  // 追加固定文本"切换悬浮窗"
            break;                                  // 跳出 switch
    }
    // 追加子命令（组合键），同样翻译成可读键名
    for (const int sub : subCommands)       // 【C++ 语法】范围 for（range-based for）：sub 依次取子命令列表中的每个键码
        parts << keyCodeToName(sub);        // 把子命令键码翻译成键名，追加到列表
    return parts.join(QStringLiteral("+")); // 【Qt】join("+")：用"+"把列表拼接成 "B+Ctrl" 形式的字符串返回
}

// ============================================================
// ControllerProfile::createDefault：创建默认配置
// ============================================================
// 返回一份全新的默认 ControllerProfile：
//   - 默认操作集「默认操作集」（Set1）：含公共层 + 10 个空操作层。
//   - 公共层（Common）：绑定常用键（空格/左右键/ESC 等）+ 视角控制，
//     不预设层切换映射，由用户通过 UI 或配置文件自行设置；
//   - 10 个操作层：各自带显示名，具体键位映射留空，由用户编辑。
// 【C++ 语法】返回值类型为类类型 ControllerProfile：按值返回整个配置对象。
ControllerProfile ControllerProfile::createDefault() {   // 实现静态工厂方法：生成默认配置
    ControllerProfile p;                                 // 创建空配置对象
    p.operationSets.append(OperationSet::createEmpty(QStringLiteral("Set1"),
                                                     QStringLiteral("默认操作集")));
    // 【Qt】QStringLiteral：编译期把字符串字面量预转为 QString，避免运行时转换开销；
    // append() 追加：创建 id="Set1"/显示名"默认操作集"的空操作集（含公共层 + 10 个空操作层）。
    p.activeOperationSetId = QStringLiteral("Set1");     // 把默认激活操作集 id 设为 "Set1"
    OperationSet& set = *p.activeSet();                  // 【C++ 语法】引用 &：对 activeSet() 返回的指针解引用并取别名，之后直接操作 set

    // ---- 公共层默认映射 ----
    // 基础常用键：供所有层共享，操作层没有映射的键会回退到这里
    // 【C++ 语法】花括号初始化 {}：KeyMapping{主动作, 子命令列表} 聚合初始化；子命令传 {} 表示空列表（QVector 默认构造）。
    set.commonLayer.buttonMappings.insert(               // 【Qt】QHash::insert(键, 值)：向公共层映射表插入 按钮->映射
        ControllerButton::A, KeyMapping{MappedAction::keyboardKey(AndroidKey::SPACE), {}});   // A 键 -> 键盘"空格"
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::B, KeyMapping{MappedAction::mouseClick(MouseButton::RIGHT), {}});   // B 键 -> 鼠标右键单击
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::X, KeyMapping{MappedAction::mouseClick(MouseButton::LEFT), {}});    // X 键 -> 鼠标左键单击
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::Y, KeyMapping{MappedAction::keyboardKey(AndroidKey::I), {}});       // Y 键 -> 键盘"I"
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::MENU, KeyMapping{MappedAction::keyboardKey(AndroidKey::ESCAPE), {}});   // MENU 键 -> 键盘"ESC"
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::OPTIONS, KeyMapping{MappedAction::keyboardKey(AndroidKey::M), {}});     // OPTIONS 键 -> 键盘"M"
    // 右摇杆按压 -> 视角控制
    set.commonLayer.buttonMappings.insert(               // 继续插入公共层映射
        ControllerButton::RIGHT_STICK_CLICK, KeyMapping{MappedAction::lookAround(), {}});    // 右摇杆按压 -> 视角控制（配合鼠标移动实现视角转向）

    return p;                                            // 返回生成的默认配置
}

File: GamepadControlerRustEgui/src/core/mapping_types.rs
// =====================================================================
// mapping_types.rs —— 映射数据模型
//
// 等效 Qt 版 MappingTypes.h（MappedAction / KeyMapping / OperationLayer /
// GlobalSettings / OperationSet / ControllerProfile）。
//
// 三层结构总览（从高到低）：
//   操作集（OperationSet）—— 最顶层容器，由 1 个公共层 + 最多 10 个
//     操作层组成；切换操作集时，其下所有操作层整体切换。
//   公共层（common_layer）—— 始终激活，优先度最低，作为兜底映射；
//   操作层（layers）—— 通过公共层的 SwitchLayer 映射按住激活、松开回退。
//  按键查询顺序（get_effective_mapping）：最后激活的操作层 -> ... -> 公共层。
// =====================================================================

use crate::core::input_types::*;
use std::collections::HashMap;

/// 单个操作集内最多操作层数
pub const K_MAX_LAYERS_PER_SET: usize = 10;

// ---------------------------------------------------------------------
// MappedAction —— 映射动作
// ---------------------------------------------------------------------
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ActionType {
    KeyboardKey,
    MouseClick,
    SwitchLayer,
    MouseMove,
    LookAround,
    MouseToggle,
    WheelUp,
    WheelDown,
    ToggleMapping,
    ToggleOnScreenKeyboard,
    ToggleOverlay,
}

// MappedAction 含 Option<String>（layer_name），不能 Copy，只 Clone。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MappedAction {
    pub r#type: ActionType,
    pub key_code: i32,
    pub mouse_button: MouseButton,
    pub layer_name: Option<String>,
}

impl MappedAction {
    pub fn keyboard_key(code: i32) -> Self {
        Self {
            r#type: ActionType::KeyboardKey,
            key_code: code,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn mouse_click(b: MouseButton) -> Self {
        Self {
            r#type: ActionType::MouseClick,
            key_code: 0,
            mouse_button: b,
            layer_name: None,
        }
    }
    pub fn mouse_toggle(b: MouseButton) -> Self {
        Self {
            r#type: ActionType::MouseToggle,
            key_code: 0,
            mouse_button: b,
            layer_name: None,
        }
    }
    pub fn wheel_up() -> Self {
        Self {
            r#type: ActionType::WheelUp,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn wheel_down() -> Self {
        Self {
            r#type: ActionType::WheelDown,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn switch_layer(name: &str) -> Self {
        Self {
            r#type: ActionType::SwitchLayer,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: Some(name.to_string()),
        }
    }
    pub fn mouse_move() -> Self {
        Self {
            r#type: ActionType::MouseMove,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn look_around() -> Self {
        Self {
            r#type: ActionType::LookAround,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn toggle_mapping() -> Self {
        Self {
            r#type: ActionType::ToggleMapping,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn toggle_on_screen_keyboard() -> Self {
        Self {
            r#type: ActionType::ToggleOnScreenKeyboard,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
    pub fn toggle_overlay() -> Self {
        Self {
            r#type: ActionType::ToggleOverlay,
            key_code: 0,
            mouse_button: MouseButton::Left,
            layer_name: None,
        }
    }
}

// ---------------------------------------------------------------------
// KeyMapping —— 单个按键映射
// ---------------------------------------------------------------------
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyMapping {
    pub action: MappedAction,
    pub sub_commands: Vec<i32>, // Android KeyCode，最多 MAX_SUB_COMMANDS 个
}

impl KeyMapping {
    pub const MAX_SUB_COMMANDS: usize = 3;

    /// 生成可读描述字符串，如 "W+ALT"
    pub fn describe(&self) -> String {
        let mut parts: Vec<String> = Vec::new();
        match self.action.r#type {
            ActionType::KeyboardKey => parts.push(key_code_to_name(self.action.key_code)),
            ActionType::MouseClick => parts.push(mouse_button_display_name(self.action.mouse_button).to_string()),
            ActionType::MouseToggle => parts.push(format!("长按{}", mouse_button_display_name(self.action.mouse_button))),
            ActionType::WheelUp => parts.push("滚轮上滚".to_string()),
            ActionType::WheelDown => parts.push("滚轮下滚".to_string()),
            ActionType::SwitchLayer => parts.push(format!("切换→{}", self.action.layer_name.as_deref().unwrap_or(""))),
            ActionType::MouseMove => parts.push("鼠标移动".to_string()),
            ActionType::LookAround => parts.push("视角控制".to_string()),
            ActionType::ToggleMapping => parts.push("切换映射".to_string()),
            ActionType::ToggleOnScreenKeyboard => parts.push("切换屏幕键盘".to_string()),
            ActionType::ToggleOverlay => parts.push("切换悬浮窗".to_string()),
        }
        for sub in &self.sub_commands {
            parts.push(key_code_to_name(*sub));
        }
        parts.join("+")
    }
}

// ---------------------------------------------------------------------
// OperationLayer —— 操作层（一组按键映射）
// ---------------------------------------------------------------------
#[derive(Debug, Clone)]
pub struct OperationLayer {
    /// 唯一标识符，固定不变（如 "Layer1"、"Common"），运行时逻辑只认 id
    pub id: String,
    /// 显示名称，用户可修改
    pub name: String,
    pub has_trigger_button: bool,
    pub trigger_button: ControllerButton,
    /// 按钮 -> 映射
    pub button_mappings: HashMap<ControllerButton, KeyMapping>,
}

impl OperationLayer {
    pub fn new(id: &str) -> Self {
        Self {
            id: id.to_string(),
            name: id.to_string(),
            has_trigger_button: false,
            trigger_button: ControllerButton::A,
            button_mappings: HashMap::new(),
        }
    }

    pub fn get_mapping(&self, b: ControllerButton) -> Option<&KeyMapping> {
        self.button_mappings.get(&b)
    }
}

// ---------------------------------------------------------------------
// GlobalSettings —— 全局设置
// ---------------------------------------------------------------------
#[derive(Debug, Clone)]
pub struct GlobalSettings {
    pub deadzone: f32,
    pub look_sensitivity: f32,
    pub cursor_speed: f32,
    pub look_smoothing: f32,
    pub look_acceleration: f32,
    pub invert_look_x: bool,
    pub invert_look_y: bool,
    pub overlay_x: i32,
    pub overlay_y: i32,
    pub overlay_scale: f64,
    pub main_window_x: i32,
    pub main_window_y: i32,
    pub release_on_foreground_change: bool,
    pub confirm_on_close: bool,
}

impl Default for GlobalSettings {
    fn default() -> Self {
        Self {
            deadzone: 0.15,
            look_sensitivity: 0.5,
            cursor_speed: 1.0,
            look_smoothing: 0.5,
            look_acceleration: 1.5,
            invert_look_x: false,
            invert_look_y: false,
            overlay_x: -1,
            overlay_y: -1,
            overlay_scale: 1.0,
            main_window_x: -1,
            main_window_y: -1,
            release_on_foreground_change: true,
            confirm_on_close: true,
        }
    }
}

// ---------------------------------------------------------------------
// OperationSet —— 操作集（最顶层容器）
// ---------------------------------------------------------------------
#[derive(Debug, Clone)]
pub struct OperationSet {
    pub id: String,
    pub name: String,
    pub common_layer: OperationLayer,
    pub layers: Vec<OperationLayer>,
}

impl OperationSet {
    /// 创建一个全新的空操作集：空公共层 + K_MAX_LAYERS_PER_SET 个空操作层
    pub fn create_empty(set_id: &str, set_name: &str) -> Self {
        let mut common = OperationLayer::new("Common");
        common.name = "Common".to_string();
        let layer_ids = [
            "Layer1", "Layer2", "Layer3", "Layer4", "Layer5",
            "Layer6", "Layer7", "Layer8", "Layer9", "Layer10",
        ];
        let layers = layer_ids
            .iter()
            .map(|id| {
                let mut l = OperationLayer::new(id);
                l.name = layer_display_name(id);
                l
            })
            .collect();
        Self {
            id: set_id.to_string(),
            name: set_name.to_string(),
            common_layer: common,
            layers,
        }
    }
}

// ---------------------------------------------------------------------
// ControllerProfile —— 完整配置
// ---------------------------------------------------------------------
#[derive(Debug, Clone)]
pub struct ControllerProfile {
    /// 操作集列表（至少 1 个）
    pub operation_sets: Vec<OperationSet>,
    /// 当前激活的操作集 id
    pub active_operation_set_id: String,
    pub global_settings: GlobalSettings,
}

impl ControllerProfile {
    /// 当前激活的操作集（无有效激活集时回退第一个；空则 None）
    pub fn active_set(&self) -> Option<&OperationSet> {
        self.operation_sets
            .iter()
            .find(|s| s.id == self.active_operation_set_id)
            .or_else(|| self.operation_sets.first())
    }

    pub fn active_set_mut(&mut self) -> Option<&mut OperationSet> {
        let id = self.active_operation_set_id.clone();
        if let Some(pos) = self.operation_sets.iter().position(|s| s.id == id) {
            Some(&mut self.operation_sets[pos])
        } else {
            self.operation_sets.first_mut()
        }
    }

    /// 当前激活操作集的公共层
    pub fn common_layer(&self) -> Option<&OperationLayer> {
        self.active_set().map(|s| &s.common_layer)
    }

    /// 当前激活操作集的公共层（可变）
    pub fn common_layer_mut(&mut self) -> Option<&mut OperationLayer> {
        self.active_set_mut().map(|s| &mut s.common_layer)
    }

    /// 当前激活操作集的操作层
    pub fn layers(&self) -> Vec<&OperationLayer> {
        match self.active_set() {
            Some(s) => s.layers.iter().collect(),
            None => Vec::new(),
        }
    }

    pub fn layers_mut(&mut self) -> Vec<&mut OperationLayer> {
        match self.active_set_mut() {
            Some(s) => s.layers.iter_mut().collect(),
            None => Vec::new(),
        }
    }

    /// 当前激活操作集的显示名
    pub fn active_operation_set_name(&self) -> String {
        self.active_set().map(|s| s.name.clone()).unwrap_or_default()
    }

    /// 按 id 设置当前激活操作集；无效 id 返回 false
    pub fn set_active_operation_set(&mut self, id: &str) -> bool {
        if self.operation_sets.iter().any(|s| s.id == id) {
            self.active_operation_set_id = id.to_string();
            true
        } else {
            false
        }
    }

    /// 生成一个不与现有操作集重复的新 id（"Set1"、"Set2"...）
    pub fn unique_operation_set_id(&self) -> String {
        let mut max = 0;
        for s in &self.operation_sets {
            if let Some(rest) = s.id.strip_prefix("Set") {
                if let Ok(n) = rest.parse::<i32>() {
                    if n > max {
                        max = n;
                    }
                }
            }
        }
        format!("Set{}", max + 1)
    }

    /// 按 id 查找层（仅当前激活操作集内，含公共层）
    pub fn find_layer(&self, id: &str) -> Option<&OperationLayer> {
        let set = self.active_set()?;
        if set.common_layer.id == id {
            return Some(&set.common_layer);
        }
        set.layers.iter().find(|l| l.id == id)
    }

    /// 由触发按键找操作层（当前激活操作集内，仅供 UI 展示）
    pub fn find_layer_by_trigger(&self, b: ControllerButton) -> Option<&OperationLayer> {
        let set = self.active_set()?;
        set.layers
            .iter()
            .find(|l| l.has_trigger_button && l.trigger_button == b)
    }

    /// 生成默认配置（WoW 预设：1 个"默认操作集"，含公共层 + 10 个操作层）
    pub fn create_default() -> Self {
        let mut p = Self {
            operation_sets: vec![OperationSet::create_empty("Set1", "默认操作集")],
            active_operation_set_id: "Set1".to_string(),
            global_settings: GlobalSettings::default(),
        };
        let common = p.common_layer_mut().expect("default set exists");
        // ---- 公共层默认映射 ----
        common.button_mappings.insert(
            ControllerButton::A,
            KeyMapping {
                action: MappedAction::keyboard_key(android_key::SPACE),
                sub_commands: vec![],
            },
        );
        common.button_mappings.insert(
            ControllerButton::B,
            KeyMapping {
                action: MappedAction::mouse_click(MouseButton::Right),
                sub_commands: vec![],
            },
        );
        common.button_mappings.insert(
            ControllerButton::X,
            KeyMapping {
                action: MappedAction::mouse_click(MouseButton::Left),
                sub_commands: vec![],
            },
        );
        common.button_mappings.insert(
            ControllerButton::Y,
            KeyMapping {
                action: MappedAction::keyboard_key(android_key::I),
                sub_commands: vec![],
            },
        );
        common.button_mappings.insert(
            ControllerButton::Menu,
            KeyMapping {
                action: MappedAction::keyboard_key(android_key::ESCAPE),
                sub_commands: vec![],
            },
        );
        common.button_mappings.insert(
            ControllerButton::Options,
            KeyMapping {
                action: MappedAction::keyboard_key(android_key::M),
                sub_commands: vec![],
            },
        );
        // 右摇杆按压 -> 视角控制
        common.button_mappings.insert(
            ControllerButton::RightStickClick,
            KeyMapping {
                action: MappedAction::look_around(),
                sub_commands: vec![],
            },
        );
        p
    }
}
